Add support for Spot VMs

JoelSpeed opened the pull request:
**What this PR does / why we need it**:

Adds a `spotVMOptions` field to the AzureMachineSpec to allow users to request a Spot VM rather than a standard on-demand VM

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #407 (Should 407 cover machinepool too?)

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add support for Spot VMs to AzureMachine
```

detiber commented on api/v1alpha3/azuremachine_types.go:
```diff
@@ -59,6 +59,18 @@ type AzureMachineSpec struct {
+
+	// SpotVMOptions allows the ability to specify the Machine should use a Spot VM
+	// +optional
+	SpotVMOptions *SpotVMOptions `json:"spotVMOptions,omitempty"`
```
There should probably be a conversion in place to backup/restore this value from annotation during conversion to avoid a lossy round trip.

CecileRobertMichon commented:
@JoelSpeed do you have plans to document this feature in the cluster API book or does it make sense to add a short doc in [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/tree/master/docs/topics) for now? 

also FYI @devigned this is probably something we'll want to bring to the vmss implementation after both PRs merge or possibly in this one if #483  merges first.

CecileRobertMichon commented:
Opened #575 to track AzureMachinePool support separately.

JoelSpeed commented:
@CecileRobertMichon I was planning to have docs in the main CAPI book eventually, but not sure when that will be done. For now I'm happy to add a short doc in this PR which can be removed later once the full CAPI docs are done

CecileRobertMichon commented:
@JoelSpeed this still needs docs + a rebase when you get a chance

JoelSpeed commented:
@CecileRobertMichon Apologies for the delay! I have rebased and added a small doc, not sure how much is appropriate to write here, I can expand more if you would like

CecileRobertMichon commented on api/v1alpha3/azuremachine_types.go:
```diff
@@ -72,6 +72,18 @@ type AzureMachineSpec struct {
+	// MaxPrice defines the maximum price the user is willing to pay for Spot VM instances
+	// +optional
+	// +kubebuilder:validation:pattern="^[0-9]+(\.[0-9]+)?$"
+	MaxPrice *string `json:"maxPrice,omitempty"`
```
Sorry I missed this earlier, why `MaxPrice` a string pointer? I think a float pointer would be more appropriate no? Maybe `*float64`

JoelSpeed replied:
Yeah it would be, I did this quite a while ago, I can't remember why I didn't do that, as that is what is described in the proposal too. Will investigate changing this

JoelSpeed replied:
I remember now why this can't be a float64. The CRD generator won't allow you to have a float64

https://github.com/kubernetes-sigs/controller-tools/blob/581f622859a03f041883454544414b1a179ebb8a/pkg/crd/schema.go#L409-L410

CecileRobertMichon replied:
I wonder if `// +kubebuilder:validation:Type=number` would be equivalent?

CecileRobertMichon commented on docs/topics/spot-vms.md:
```diff
@@ -0,0 +1,52 @@
+spec:
+  template:
+    spotMarketOptions:
+      maxPrice: "0.04" Price in USD per hour (up to 5 decimal places)
```
is ` Price in USD per hour (up to 5 decimal places)` a comment? it's showing as part of the string right now

JoelSpeed replied:
Yeah that was meant as a comment, fixed

CecileRobertMichon commented on docs/topics/spot-vms.md:
````diff
@@ -0,0 +1,52 @@
+```
+
+You may also add a `maxPrice` to the options to limit the maximum spend for the
+instance. It is however, recommend **not** to set a `maxPrice` as Azure will
````
```suggestion
instance. It is however, recommended **not** to set a `maxPrice` as Azure will
```

CecileRobertMichon commented on docs/topics/spot-vms.md:
```diff
@@ -0,0 +1,52 @@
+apiVersion: exp.infrastructure.cluster.x-k8s.io/v1alpha3
+kind: AzureMachineTemplate
+metadata:
+  name: capz-mp-0
```
mp stands for machine pool usually, this should be `md`

CecileRobertMichon commented on docs/topics/spot-vms.md:
````diff
@@ -0,0 +1,52 @@
+to your `AzureMachineTemplate`:
+
+```yaml
+apiVersion: exp.infrastructure.cluster.x-k8s.io/v1alpha3
````
the `exp` apiVersion is for MachinePools

CecileRobertMichon commented on docs/topics/spot-vms.md:
```diff
@@ -0,0 +1,52 @@
+      osType: Linux
+    sshPublicKey: ${YOUR_SSH_PUB_KEY}
+    vmSize: Standard_D2s_v3
+    spotMarketOptions: {}
```
Wouldn't it make more sense to have spotMarketOptions.Enabled set to true or false? Having an empty field to "enable" a feature seems a little strange... I realize this is an API question but looking at the example is what made me realize that was the way of enabling it without maxPrice

JoelSpeed replied:
This was inspired by the `emptyDir: {}` volume from core Kubernetes. This was documented in the proposal to cluster-api and the same approach was intended to be taken for the AWS provider too. If you think we should change this, we should have a discussion about changing the proposal and seeing what the AWS folks think as well, ideally there would be consistency between the two providers IMO

CecileRobertMichon replied:
@devigned @detiber any thoughts on this one?

devigned replied:
That looks really odd.

I'm not sure what is worse, having to set enabled and possibly a max price, which seem redundant, or having to specify an empty.

devigned replied:
It seems to be part of the idiom: https://kubernetes.io/docs/concepts/storage/volumes/#example-pod. Anyone have any better ideas?

CecileRobertMichon replied:
the `emptyDir` volume feels slightly different to me because it does point to something being empty. SpotVMs is a feature you can enable/disable, with optional options, like NTP https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/api/v1alpha2/kubeadmbootstrapconfig_types.go#L197 so I would prefer something like:

```
// Spot defines the options relevant to running the Machine on Spot VMs
type Spot struct {
	// MaxPrice defines the maximum price the user is willing to pay for Spot VM instances
	// +optional
	// +kubebuilder:validation:Type=number
	MaxPrice *string `json:"maxPrice,omitempty"`

	// Enabled specifies whether Spot should be enabled
	// +optional
	Enabled *bool `json:"enabled,omitempty"`
}
```

Would be good to get CAPA folks to chime in as well since we want to be consistent across providers.

devigned replied:
@kubernetes-sigs/cluster-api-provider-aws-maintainers you all have any thoughts on how CAPA is thinking about approaching spot pricing? I did a quick search through the issues and didn't see anything.

detiber replied:
I don't think the empty struct vs nil bothers me much if at all, since in the default case a user just wouldn't specify the field and would not get a Spot instance, only when they explicitly initialize/set the field do they get a Spot instance.

To me it would likely feel a bit odder if you were to explicitly configure the optional fields for the SpotVMOptions but omit setting Enabled to true and end up with a non-spot VM.

CecileRobertMichon replied:
Thanks for chiming it @detiber. I'm okay with going with empty options if that's the general consensus and since that was the original agreed-upon design. @devigned wdyt?

devigned replied:
👍 Let's do this!

@JoelSpeed, do you want to continue this PR? Since it's been a minute, just wanted to see if you were still interested in carrying this over the finish line. If you don't have time, I'm happy to help.

JoelSpeed replied:
Yep, happy to continue, will rebase tomorrow

CecileRobertMichon commented on api/v1alpha3/azuremachine_types.go:
```diff
@@ -72,6 +72,18 @@ type AzureMachineSpec struct {
+type SpotVMOptions struct {
+	// MaxPrice defines the maximum price the user is willing to pay for Spot VM instances
+	// +optional
+	// +kubebuilder:validation:pattern="^[0-9]+(\.[0-9]+)?$"
```
I wonder if // +kubebuilder:validation:Type=number would be equivalent?

JoelSpeed replied:
I added this and it works! Means that the `maxPrice` must be specified without quotes now, so I've updated the doc too, is that preferred? Seems a shame we can't have both

CecileRobertMichon replied:
I think without quotes makes more sense anyways since to the user it's a number, not a string

JoelSpeed commented on docs/topics/spot-vms.md:
```diff
@@ -0,0 +1,52 @@
+      osType: Linux
+    sshPublicKey: ${YOUR_SSH_PUB_KEY}
+    vmSize: Standard_D2s_v3
+    spotMarketOptions: {}
```
This should be `spotVMOptions`, I'm confusing Azure and AWS

JoelSpeed commented:
@CecileRobertMichon Is there anything more you want me to do with this? Did we want to raise the question about the `spotVMOptions: {}` emptiness being the default usage? I don't mind it but if you think there's a potential to improve the UX then we should discuss

CecileRobertMichon commented:
@JoelSpeed the only two things left for me are 1) merge https://github.com/kubernetes-sigs/cluster-api/pull/3063 and 2) resolve the question on `spotVMOptions: {}` usage. I'm not a big fan of having an empty "options" field to enable it but I've asked for second opinions on the thread above since this will likely affect CAPA as well.

nader-ziada commented:
this looks good to me and ready to be merged, but asking @CecileRobertMichon if she has other comments since she has more context
